rust/agama-files/src/runner.rs

```diff
@@ -0,0 +1,365 @@
+    command::run_with_retry,
+    progress,
+    question::{self, ask_question},
+};
+use tokio::process;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error("The script failed")]
+    Script(Output),
+}
+
+/// Implements the logic to run a script.
+///
+/// It takes care of running the script, reporting errors (and asking whether to retry) and write
+/// the logs.
+pub struct ScriptsRunner {
+    progress: Handler<progress::Service>,
+    questions: Handler<question::Service>,
+    install_dir: PathBuf,
+    workdir: PathBuf,
+}
+
+impl ScriptsRunner {
+    /// Creates a new runner.
+    ///
+    /// * `install_dir`: directory where the system is being installed. It is relevant for
+    ///   chrooted scripts.
+    /// * `workdir`: scripts work directory.
+    /// * `progress`: handler to report the progress.
+    /// * `questions`: handler to interact with the user.
+    pub fn new<P: AsRef<Path>>(
+        install_dir: P,
+        workdir: P,
+        progress: Handler<progress::Service>,
+        questions: Handler<question::Service>,
+    ) -> Self {
+        Self {
+            progress,
+            questions,
+            install_dir: install_dir.as_ref().to_path_buf(),
+            workdir: workdir.as_ref().to_path_buf(),
+        }
+    }
+
+    /// Runs the given scripts.
+    ///
+    /// It runs each script. If something goes wrong, it reports the problem to the user through
+    /// the questions mechanism.
+    ///
+    /// * `scripts`: scripts to run.
+    pub async fn run(&self, scripts: &[&Script]) -> Result<(), Error> {
+        self.start_progress(scripts);
+
+        for script in scripts {
+            _ = self
+                .progress
+                .cast(progress::message::Next::new(Scope::Files));
+            self.run_script(script).await;
+        }
+
+        _ = self
+            .progress
+            .cast(progress::message::Finish::new(Scope::Files));
+        Ok(())
+    }
+
+    /// Runs the script.
+    ///
+    /// If the script fails, it asks the user whether it should try again.
+    async fn run_script(&self, script: &Script) {
+        loop {
+            let path = self
+                .workdir
+                .join(script.group().to_string())
+                .join(script.name());
+
+            let Err(error) = self.run_command(&path, script.chroot()).await else {
+                return;
+            };
+
+            if !self.should_retry(&script, error).await {
+                return;
+            }
+        }
+    }
+
+    ///  Asks the user whether it should try to run the script again.
+    async fn should_retry(&self, script: &Script, error: Error) -> bool {
+        let text = format!(
+            "Running the script '{}' failed. Do you want to try again?",
+            script.name()
+        );
+        let mut question = QuestionSpec::new(&text, "scripts.retry").with_yes_no_actions();
```

Comment:
do we have that type of question already documented? If not, please write it down to some follow up, to update documentation when new api will be merged.

Reply:
that maybe open question if we should document that it is only for 16.1 and newer, but it is topic for that documentation PR